File: ios/UnityNativeModule.m
//
//  UnityNativeModule.m
//  RNUnityView
//
//  Created by xzper on 2018/12/13.
//  Copyright © 2018 xzper. All rights reserved.
//

#import "UnityNativeModule.h"

@implementation UnityNativeModule

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE(UnityNativeModule);

- (id)init
{
    NSLog(@"UUUUUUUUUUUU In UnityNativeModule.init");
    self = [super init];
    if (self) {
        NSLog(@"UUUUUUUUUUUU In UnityNativeModule.init, self is true so about to call addUnityEventListener");
        [UnityUtils addUnityEventListener:self];
        NSLog(@"UUUUUUUUUUUU In UnityNativeModule.init, self is true, called addUnityEventListener");
    }
    return self;
}

- (NSArray<NSString *> *)supportedEvents
{
        NSLog(@"UUUUUUUUUUUU In UnityNativeModule.supportedEvents");
    return @[@"onUnityMessage"];
}

+ (BOOL)requiresMainQueueSetup
{
        NSLog(@"UUUUUUUUUUUU In UnityNativeModule.requiresMainQueueSetup");
    return YES;
}

RCT_EXPORT_METHOD(isReady:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
        NSLog(@"UUUUUUUUUUUU In UnityNativeModule.isReady");
    resolve(@([UnityUtils isUnityReady]));
}

RCT_EXPORT_METHOD(createUnity:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    NSLog(@"UUUUUUUUUUUU In UnityNativeModule.createUnity, about to call createPlayer");
    [UnityUtils createPlayer:^{
     NSLog(@"UUUUUUUUUUUU In UnityNativeModule.createUnity, in createPlayer callback, about to call resolve");
       resolve(@(YES));
    }];
}

RCT_EXPORT_METHOD(postMessage:(NSString *)gameObject methodName:(NSString *)methodName message:(NSString *)message)
{
    NSLog(@"UUUUUUUUUUUU In UnityNativeModule.postMessage about to call UnityPostMessage");
    UnityPostMessage(gameObject, methodName, message);
    NSLog(@"UUUUUUUUUUUU In UnityNativeModule.postMessage, called UnityPostMessage");
}

RCT_EXPORT_METHOD(pause)
{
    NSLog(@"UUUUUUUUUUUU In UnityNativeModule.pause");
    UnityPauseCommand();
}

RCT_EXPORT_METHOD(resume)
{
     NSLog(@"UUUUUUUUUUUU In UnityNativeModule.resume");
   UnityResumeCommand();
}

- (void)onMessage:(NSString *)message {
     NSLog(@"UUUUUUUUUUUU In UnityNativeModule.onMessage %@", message);
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
    [_bridge.eventDispatcher sendDeviceEventWithName:@"onUnityMessage"
                                                body:message];
#pragma clang diagnostic pop
}

@end

File: android/src/main/java/com/reactnative/unity/view/UnityNativeModule.java
package com.reactnative.unity.view;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.DeviceEventManagerModule;

public class UnityNativeModule extends ReactContextBaseJavaModule implements UnityEventListener {

    public UnityNativeModule(ReactApplicationContext reactContext) {
        super(reactContext);
        UnityUtils.addUnityEventListener(this);
    }

    @Override
    public String getName() {
        return "UnityNativeModule";
    }

    @ReactMethod
    public void isReady(Promise promise) {
        promise.resolve(UnityUtils.isUnityReady());
    }

    @ReactMethod
    public void createUnity(final Promise promise) {
         System.out.println("UUUUUUUUUUUU In UnityNativeModule.createUnity");
       UnityUtils.createPlayer(getCurrentActivity(), new UnityUtils.CreateCallback() {
            @Override
            public void onReady() {
                promise.resolve(true);
            }
        });
    }

    @ReactMethod
    public void postMessage(String gameObject, String methodName, String message) {
          System.out.println("UUUUUUUUUUUU In UnityNativeModule.postMessage: " + message);
       UnityUtils.postMessage(gameObject, methodName, message);
    }

    @ReactMethod
    public void pause() {
         System.out.println("UUUUUUUUUUUU In UnityNativeModule.pause");
        UnityUtils.pause();
    }

    @ReactMethod
    public void resume() {
         System.out.println("UUUUUUUUUUUU In UnityNativeModule.resume");
        UnityUtils.resume();
    }

    @Override
    public void onMessage(String message) {
          System.out.println("UUUUUUUUUUUU In UnityNativeModule.onMessage: " + message);
        ReactContext context = getReactApplicationContext();
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("onUnityMessage", message);
    }
}
